### PR TITLE
Improvement of the HPO wrapper and Handling Training Configuration & Introduction of SpilloverNoiseQuantumEnvironment 

### PR DESCRIPTION
**Pull request by lukasvoss**
This PR extends the current capabilities on how to use the hyperparameter optimization (HPO) wrapper, introduces a new approach to how a user sets up his RL training agent and adds a new dedicated class for dealing with the Pulse Spillover Noise use case.

**HPO Wrapper**
- Can be used for all calibration tasks and will handle the individual cases based on the provided input configurations
- Automatically adjusts the number repetitions of the target gate in the spillover noise use case for a given target fidelity and noise profile $(\phi, \gamma)$

**Improved Handling of how to specify the training parameters**
- All RL training-related settings are now to be set in dedicated dataclasses: See the file _hpo_training_config.py_ that both has the associated dataclasses for an hpo-case and normal training case
- The user can set 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Between two types of training constraints (total number of updates or a maximum associated hardware runtime) 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - logging behaviour of the agent, learning rate annealing, live plotting, target fidelities, set a threshold that stops the training if the standard deviations of the actions are below it, how large the lookback of the sliding window approach is (that records the surpassing of the provided target fidelities), and beyond

**Dedicated class PulseSpilloverQuantumEnvironment for this use case**
- Inherits from the `ContextAwareQuantumEnvironmentV2` and takes the $(\phi, \gamma)$ tuple as additional argument
- Class has a function that sets the _n_reps_ of the environment to the number needed amplify the noise profile below the lowest target fidelity value provided to the function

- There is a ReadMe file for the Pulse Spillover Noise use case

### CODE REVIEW

File: context_aware_quantum_environment.py [hunk 1]
```diff
@@ -1212,3 +1230,149 @@ def clear_history(self) -> None:
         """Reset all counters related to training"""
         super().clear_history()
         self.circuit_fidelity_history.clear()
+
+    def update_circuit_fidelity_history(self, output_states, baseline_circ):
```
Comment (arthurostrauss) [inline context_aware_quantum_environment.py:1234]: I removed this method in the latest file because it was not adding anything useful.

File: context_aware_quantum_environment.py [hunk 2]
```diff
@@ -1212,3 +1230,149 @@ def clear_history(self) -> None:
...
+        circuit_fidelities = [
+            state_fidelity(state, Statevector(baseline_circ)) for state in output_states
+        ]
+        self.circuit_fidelity_history.append(np.mean(circuit_fidelities))
+        return circuit_fidelities
+    
+
+class SpilloverNoiseQuantumEnvironment(ContextAwareQuantumEnvironmentV2):
```
Comment (arthurostrauss) [inline context_aware_quantum_environment.py:1242]: Move your custom class in a separate Python file. This will be in your custom use case folder.

File: hyperparameter_optimization.py [hunk 3]
```diff
@@ -219,64 +342,112 @@ def _logging_result(self, study: optuna.study.Study, start_time: float):
...
+    def training_mode(self):
+        return self.training_config.training_mode
 
     @property
-    def target_gate(self):
+    def phi(self):
+        return self.q_env.unwrapped.phi_gamma_tuple[0] if isinstance(self.q_env.unwrapped, SpilloverNoiseQuantumEnvironment) else None
+    
+    @property
+    def gamma(self):
```
Comment (arthurostrauss) [inline hyperparameter_optimization.py:430]: The HPO should not have such property. Are you still using it in the main code? If yes, then we should find a workaround, if not you can just remove it.

**Review by arthurostrauss — CHANGES_REQUESTED** [review]
Excellent job. Just one last thing to address in your HPO, it still carries a property reminiscent of your custom use case.

File: hyperparameter_optimization.py (hunk 3 above)
Comment (lukasvoss) [reply hyperparameter_optimization.py]: Good catch, you're right. I made phi and gamma a property of the `SpilloverNoiseQuantumEnvironment` that we can access from the HPO in the relevant lines that already check that our env is of this instance.

File: context_aware_quantum_environment.py (hunk 1 above)
Comment (lukasvoss) [reply context_aware_quantum_environment.py]: Removed it. Did I see it right, that this function was a duplicate?

File: context_aware_quantum_environment.py (hunk 2 above)
Comment (lukasvoss) [reply context_aware_quantum_environment.py]: Moved it.

File: hyperparameter_optimization.py (hunk 3 above)
Comment (arthurostrauss) [reply hyperparameter_optimization.py]: We should therefore discuss where in the HPO do you need this information, as I still believe it should not carry any information about the subclasses you have created for a custom use case. I seem to have located references to phi and gamma only for logging options. What I suggest you therefore do is that you proceed to a similar fashion to what you did with the modify_env_parameters, you create a method in the base class rendering your environment when summoned for printing/logging info, and you override it in your custom environment to log the info about gamma and phi. This way, you avoid yourself the trouble of this if/elif statement for each possible subclass of the QEnv you might do in the future.

File: hyperparameter_optimization.py [hunk 4]
```diff
@@ -131,7 +131,7 @@ def _log_training_parameters(self):
         logging.warning("Parameters:")
         if self.training_mode == 'spillover_noise_use_case' and isinstance(self.q_env.unwrapped, SpilloverNoiseQuantumEnvironment):
             param_str = "phi: {}pi; gamma: {}; Baseline Fidelity: {}; ".format(
```
Comment (arthurostrauss) [inline hyperparameter_optimization.py:133]: So this is the line I'm talking about. The idea is good, but the method for rendering this string param should be in the environment. There is already a method call __repr__ which returns a string for printing info about the env. You could choose to override this one if relevant, or just create your custom one that you call here in the base class.

File: hyperparameter_optimization.py [hunk 5]
```diff
@@ -264,7 +264,7 @@ def _calculate_custom_cost(
     def _generate_filename(self):
         """Generate the file name where the best configuration will be saved."""
         if self.training_mode == 'spillover_noise_use_case' and isinstance(self.q_env.unwrapped, SpilloverNoiseQuantumEnvironment):
-            start_file_name = f"phi-{self.phi/np.pi}pi_gamma-{round(self.gamma, 2)}"
+            start_file_name = f"phi-{self.q_env.unwrapped.phi/np.pi}pi_gamma-{round(self.q_env.unwrapped.gamma, 2)}"
```
Comment (arthurostrauss) [inline hyperparameter_optimization.py:267]: I would definitely add a _label method to the environment that I call here. The label would be this name you want to give or an updated version of this.

**Review by arthurostrauss — CHANGES_REQUESTED** [review]
Nearly there,  the last comments I have are for really clearing your HPO code to any specific dependency to use cases that could be multiplied in the future.

File: context_aware_quantum_environment.py [hunk 6]
```diff
@@ -6,7 +6,10 @@
 Created on 26/06/2023
 """
 
+from abc import abstractmethod
```
Comment (arthurostrauss) [inline context_aware_quantum_environment.py:9]: Check that you don't have unused imports here.

**Review by arthurostrauss — CHANGES_REQUESTED** [review]
There are still a few traces of your custom use case in your HPO. Other small changes requested but the core is there.

File: ppoV2.py [hunk 7]
```diff
@@ -420,7 +426,7 @@ def take_step(
     actions[step] = action
     logprobs[step] = logprob
 
-    next_obs, reward, terminated, truncated, infos = env.step(action.cpu().numpy())
+    next_obs, reward, terminated, truncated, infos = env.unwrapped.step(action.cpu().numpy())
```
Comment (arthurostrauss) [inline ppoV2.py:429]: No on the contrary, you want to use the step function of the wrapped environment (to enable the RescaleAction and ClipAction). Overall, remove the unwrapped when the attribute you're looking for is actually belonging to the Wrapper (observation space, action space, step)

File: ppoV2.py [hunk 8]
```diff
@@ -847,6 +761,7 @@ def __init__(
         Returns:
             train (function): The training function for the PPO algorithm.
         """
+        self.env.unwrapped.clear_history()
```
Comment (arthurostrauss) [inline ppoV2.py:764]: I'm not sure to see why this is necessary here. Perhaps you want to clear the history in the frame of the HPO, but a PPO could act on an environment that already has a history in principle

File: ppoV2.py [hunk 9]
```diff
@@ -237,15 +241,15 @@ def initialize_networks(
 
     """
     actor_net = ActorNetwork(
-        env.observation_space,
+        env.unwrapped.observation_space,
```
Comment (arthurostrauss) [inline ppoV2.py:244]: Remove unwrapped.

File: ppoV2.py [hunk 10]
```diff
@@ -237,15 +241,15 @@ def initialize_networks(
...
         hidden_units,
         n_actions,
         activation_functions,
         include_critic,
         chkpt_dir,
     )
     critic_net = CriticNetwork(
-        env.observation_space, hidden_units, activation_functions, chkpt_dir_critic
+        env.unwrapped.observation_space, hidden_units, activation_functions, chkpt_dir_critic
```
Comment (arthurostrauss) [inline ppoV2.py:252]: Remove unwrapped.

**Review by arthurostrauss — CHANGES_REQUESTED** [review]

**Review by arthurostrauss — APPROVED** [review]
